src/render.ts

```diff
@@ -0,0 +1,70 @@
+type Props = Record<string, any>;
+
+export type VNode = {
+  tag: keyof HTMLElementTagNameMap | ((props: Props) => VNode);
+  props: Props;
+  children: (VNode | string)[];
+};
+
+export function createDOM(node: VNode | string): Node {
+  if (typeof node === "string") {
+    return document.createTextNode(node);
+  }
+
+  if (typeof node.tag === "function") {
+    const componentVNode = node.tag(node.props);
+    return createDOM(componentVNode);
+  }
+
+  const element = document.createElement(node.tag);
+
+  Object.entries(node.props).forEach(([name, value]) => {
+    const isClassAttribute = name === "className";
+    if (isClassAttribute) {
+      return element.setAttribute("class", value);
+    }
+
+    const isEventHandler = name.startsWith("on");
+    if (isEventHandler) {
+      const eventType = name.toLowerCase().substring(2);
+      return element.addEventListener(eventType, value);
+    }
+
+    element.setAttribute(name, value);
+  });
+
+  node.children.map(createDOM).forEach(element.appendChild.bind(element));
+
+  return element;
+}
+
+export function createElement(
+  tag: keyof HTMLElementTagNameMap | ((props: Props) => VNode),
+  props: Props,
+  ...children: (VNode | string)[]
+) {
+  props = props || {};
```

Comment:
Props 타입은 객체라 Call by Reference로 평가될텐데, 그러면 `createElement` 함수가 호출되는 것만으로 함수 외부의 값이 변경될 가능성이 있을 것 같아요.
그래서 함수의 인자를 직접 수정하는 것은 안티패턴이기는 합니다 ㅎㅎ

차라리 새로 변수를 선언해주는 것이 좋을 것 같고, 이렇게 nullish한 값을 예외처리할 때는 or 연산자가 아닌

```suggestion
  props = props ?? {};
```
nullish coalescing operator를 사용하는 것이 바람직할 것 같습니다

Reply:
변수로 선언하고 ?? 연산자를 사용하는 것으로 변경했습니다 !